### PR TITLE
Adding Quicksort

### LINKED ISSUE #273: Quick Sort
#### Description of the problem
I suggest adding quicksort algorithm in linear data structures
i.e linear_data_structures.algorithms.quick_sort()

There are some sorting algorithms already there, with same nlogn complexity, but since most inbuilt sorting algorithms are implemented using quicksort, we should also add quick sort

### PR DESCRIPTION
**Pull request by Arvind-raj06**
#### References to other Issues or PRs or Relevant literature
Fixes #273. See
https://github.com/codezonediitj/pydatastructs/issues/273 

#### Brief description of what is fixed or changed
This code fixes the quick sort issue on iterative implementation as reference with the site https://www.tutorialspoint.com/python-program-for-iterative-quick-sort

#### Other comments
Hope this code works fine and if there is any issue ping me, yes this PR also included the SWOC participation

### CODE REVIEW

File: pydatastructs/linear_data_structures/algorithms.py [hunk 1]
```diff
@@ -620,6 +621,95 @@ def swap(i, j):
...
+        for j in range(low , high):
+            if _comp(array[j], x, comp) is True:
+                i = i + 1
+                array[i], array[j] = array[j], array[i]
+        array[i + 1], array[high] = array[high], array[i + 1]
+        return (i + 1)
+
+    lower = kwargs.get('start', 0)
+    upper = kwargs.get('end', len(array) - 1)
+    stack, top = [0] * (upper - lower + 1), -1
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:690]: There is a `Stack` class (see below).
https://github.com/codezonediitj/pydatastructs/blob/78d8254156e6984ff556eb16c468cb53ab132a54/pydatastructs/miscellaneous_data_structures/stack.py#L9

Try using the above as stack. 

P.S.  The iterative version seems nice.

Comment (Arvind-raj06) [reply pydatastructs/linear_data_structures/algorithms.py:690]: Cool I will use the stack there in the program ^ ^

**Review by plazzy99 — APPROVED** [review]
LGTM

**Comment by Arvind-raj06** [comment]
@czgdp1807 Yep done

**Review by plazzy99 — APPROVED** [review]

File: pydatastructs/linear_data_structures/algorithms.py [hunk 2]
```diff
@@ -624,3 +625,89 @@ def swap(i, j):
...
+        return (i + 1)
+
+    lower = kwargs.get('start', 0)
+    upper = kwargs.get('end', len(array) - 1)
+    stack, top = Stack(), -1
+
+    stack.push(lower)
+    stack.push(upper)
+
+    while stack.is_empty==False:
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:699]:
```suggestion
    while stack.is_empty is False:
```

**Comment by czgdp1807** [comment]
There is one thing missing in the API which is the acceptance of a function which should select the pivot with respect to which partition should be done. Currently, `high` is the pivot but what if the user wants to select randomly? 
`kwargs` should also accept, `pick_pivot_element` which should be the method provided by the user. It should be optional and by default will pick `high`. `pick_pivot_element` will be passed to `partition` and will always be called as, `pick_pivot_element(low, high, array)`.

In addition to the above, please do not use Github as your editor as any commit made on Github triggers Travis CI tests leading to wastage of resources. You may follow the steps below to pull the changes to your system, running tests locally until they pass and then push here.
Assuming that you have git and pip (for python3) installed
1. `pip install pytest` (You may need to use `sudo` to provide write access to `pip`)
2. `pip install pytest-cov` (You may need to use `sudo` to provide write access to `pip`)
3. `git clone https://github.com/codezonediitj/pydatastructs`
4. `python path/to/pydatastructs/setup.py develop` (You may need to provide `sudo` access)
4. `git remote add origin_user  https://github.com/Arvind-raj06/pydatastructs/`
5. `git pull origin_user "Let'scode"`
6. Do your changes
7. `python -m pytest --doctest-modules --cov=./ --cov-report=xml -s` (under `pydatastructs`)
8. If tests pass, push using, `git push origin_user "Let'scode"`. Otherwise go to 6 and keep on repeating until tests pass.

**Comment by Arvind-raj06** [comment]
> There is one thing missing in the API which is the acceptance of a function which should select the pivot with respect to which partition should be done. Currently, `high` is the pivot but what if the user wants to select randomly?
> `kwargs` should also accept, `pick_pivot_element` which should be the method provided by the user. It should be optional and by default will pick `high`. `pick_pivot_element` will be passed to `partition` and will always be called as, `pick_pivot_element(low, high, array)`.
> 
> In addition to the above, please do not use Github as your editor as any commit made on Github triggers Travis CI tests leading to wastage of resources. You may follow the steps below to pull the changes to your system, running tests locally until they pass and then push here.
> Assuming that you have git and pip (for python3) installed
> 
> 1. `pip install pytest` (You may need to use `sudo` to provide write access to `pip`)
> 2. `pip install pytest-cov` (You may need to use `sudo` to provide write access to `pip`)
> 3. `git clone https://github.com/codezonediitj/pydatastructs`
> 4. `python path/to/pydatastructs/setup.py develop` (You may need to provide `sudo` access)
> 5. `git remote add origin_user  https://github.com/Arvind-raj06/pydatastructs/`
> 6. `git pull origin_user "Let'scode"`
> 7. Do your changes
> 8. `python -m pytest --doctest-modules --cov=./ --cov-report=xml -s` (under `pydatastructs`)
> 9. If tests pass, push using, `git push origin_user "Let'scode"`. Otherwise go to 6 and keep on repeating until tests pass.

Yes sure I had a copy but I thought using travis would be easier, ok now onwards I will do it local system

**Pull request merged**